### ISSUE TITLE
GHC-internal modules in `base`

### ISSUE THREAD

**Issue by bgamari** [haskell/core-libraries-committee#146]
(The proposal eventually approved in this thread is https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1591871779 — Bodigrim, Sep 2023)

## 1. Background

Currently the `base` package exposes many internal implementation details of the implementation `base` functionality. By "internal implementation details" we mean functions and data types that are part of GHC's realisation of some exposed function, but which were never intended to be directly used by clients of the `base` library. For instance, the `GHC.Base.mapFB` function is a necessary exposed part of the fusion framework for `map` but which GHC's authors never intended users to call.

This lack of clarity is bad for several reasons:

* Users have no way to know which functions are part of the "intended, stable API" and which are part of the "internal, implementation details".  Consequently, they may accidentally rely on the latter; they simply have no way to tell.

* GHC's developers are hampered in modifying the implementation because too much is exposed.  This imposes a high backward-compatibility burden, one that is an accident of history.

This status quo leaves much to be desired: users tend to rely on any interface available to them and therefore GHC developers are susceptible to breaking users when changing implementation details within `base`. On the other hand, there is a clear need to be able to iterate on the implementation of GHC and its `base` library: fixing compiler bugs may require the introduction of new internal yet exposed primitives (c.f. the [changes][unsafeCoerce] made in the implementation of `unsafeCoerce` in GHC 9.0) and improving runtime performance may require changes in the types of exposed internal implementation (c.f. [GHC #22946][T22946]).

These difficulties are discussed in [CLC #105][CLC105].

## 2. Proposal

We propose to classify the modules of `base` into three groups:

* **Hidden**: these are simply the existing non-exposed modules (`other-module` in Cabal terms). No change here.

* **External**: these modules comprise the public API of `base`.
  * The GHC team makes strenuous efforts to maintain the stability of this API
  * Changes to the API require a major version bump.
  * Changes require agreement of the CLC
  * Modules appear in the `exposed-modules` Cabal section

* **Internal**: these modules are part of the internal implementation of `base` functions.
  * The GHC team makes no effort to maintain the stability of this API
  * Changes to the API require only a minor version bump.
  * Changes do not require agreement of the CLC
  * Appear in the `exposed-modules` Cabal section

As of today, all modules are either Hidden or External; the [CLC policy][CLC105] is that the API of all exposed modules is subject to CLC review.

The main payload of this proposal is

* To establish the concept of an Internal module (exposed, but internal to GHC)
* To outline a process for incrementally moving modules from External to Internal.

### 2.1 Codifying the Internal vs External split

Our proposal is simply to *declare* whether a module is Internal or External, using some out-of-band mecanism like a publicly visible list.

However, future reorganizations (notably [HF tech propoosal #47][tech-prop-47]) might split `base` into two packages:
* `ghc-base`, all of whose exposed modules are Internal.
* `base`, all of whose exposed modules are External.

That would codify the distinction between Internal and External, which would be a Good Thing. But the burden of *this* proposal is simply to *make* that distinction in the first place, and start a dialogue about which modules belong in each category.

Incidentally, the `Stability` Haddock field of a module is not the same as Internal vs External distinction.  A module could be External (i.e. designed for external callers), and yet experimental and not yet stable.  That seems to be the intended purpose of the `Stability` field, although it is not well describe anywhere (please tell us there is a good specification).

We propose to document internal modules via a yet-to-be-named Haddock field.

### 2.2 Module by module summary

To make the discussion concrete, we have [characterized][sheet] each of the exposed modules in the `GHC.*` namespace along three axes:

 * **Proposed status**: one of External or Internal
 * **Real-world usage**: how many times each module is imported by end-user code.  This data was gathered from Hackage.
 * **Stability risk**: our subjective judgement about whether the module exposes internal details that may be subject to change in future compiler releases

These findings, along with the stability indicated by the modules' `Stability` Haddock field, are summarized in [this spreadsheet][sheet]. We then used these assessments to define an action plan (seen in the "Action" column) which will bring us closer to the goal of clearly delineating the stable interface of `base`. We do not intend to pursue this plan as one atomic change; rather, we intend for this plan to be an aspiration which we will iteratively approach over the course of the coming years, largely driven by the needs of the GHC developers.

The proposed actions fall into a few broad buckets:

 * *Internalize*, which denotes the GHC developers' intent to in the future open a CLC proposal to move the module from External to Internal.
 * *Hide*, which denotes the GHC developers' intent to in the future open a CLC proposal to remove the module from External to Hidden.
 * *Stabilize*, where the module is declared to be stable and reaffirmed as External

In the sections below we will discuss some of the reasoning behind these proposed actions and draw attention to some open questions.

## 3. The question of `GHC.Exts`

Historically `GHC.Exts` has been the primary entry-point for users wanting access to all of the primitives that GHC exposes (e.g. primitive types, operations, and other magic). This widely-used module poses a conundrum since, while many of these details are quite stable (e.g. `Int#`), a few others truly are exposing implementation details which cannot be safely used in a GHC-version-agnostic way (e.g. `mkApUpd0#`, `unpackClosure#`, `threadStatus#`). There are at least two ways by which this might be addressed:

 * Export only the subset of primops that we can stabilize (e.g. things like `Int#`, `Weak#`, `newArray#`, etc.) in `GHC.Exts`, leaving the rest to only be exposed via `GHC.Prim` (which should not be used by end-users), or
 * Declare the entirety of `GHC.Exts` to be unstable and export the stable subset from another namespace (e.g. `Word#` and its operations could be exposed by `GHC.Unboxed.Word`)

## 4. Non-normative interfaces

Several interfaces exposed by `base` intentionally reflect internal details of GHC's implementation and, by their nature, should change to reflect changes in the underlying implementation. Here we call such interfaces "non-normative" as they are defined not by a specification of desired Haskell interfaces but rather by the system that they reflect.

One such module is `GHC.Stats`, which allows the user to reflect on various statistics about the operation of the runtime system. If the runtime system were to change (e.g. by adding a new phase of garbage collection), users would expect the module to change as well. For this reason, we mark such non-normative interfaces as "internal".

[tech-prop-47]: https://github.com/haskellfoundation/tech-proposals/pull/47
[T22946]: https://gitlab.haskell.org/ghc/ghc/-/issues/22946
[CLC105]: https://github.com/haskell/core-libraries-committee/issues/105
[sheet]: https://docs.google.com/spreadsheets/d/1WmyYLbJIMk9Q-vK4No5qvKIIdIZwhhFFlw6iVWd1xNQ/edit#gid=1315971213
[CLC134]: https://github.com/haskell/core-libraries-committee/issues/134
[unsafeCoerce]: https://gitlab.haskell.org/ghc/ghc/-/blob/master/libraries/base/Unsafe/Coerce.hs

**Comment by tomjaguarpaw** [comment]
> We propose to divide the modules of base into three groups

`s/divide/classify/`, perhaps? "divide" makes it sound like you're intending to carry out some operation on the source repo, but I think you mean just "publish a classifcation". In the subsequent section you use "declare", and explain that the "division" is done out-of-band. I think it might be wise to foreshadow this by changing "divide" to "classify", to avoid misdirecting the reader.

**Comment by phadej** [comment]
There are two axis of distinction between external and internal proposed:

1. Changes do or do not require agreement of the CLC
2. Changes to the API require a major or a minor version bump.

**I stricly disagree with the second point.** If change is breaking, than it requires major version bump. I don't care whether the module is marked as "implementation specific internal". If you expose it, than it's part of public API.
If you want to fix that issue, introduce a new package (`ghc-base`) and version it properly. Then if `base` doesn't re-export affected stuff from `ghc-base`, it can stay at the same major version.

And as far as I understand, it's the first point which causes most friction. GHC devs need to change something in GHC specific stuff, but have to go through CLC. And that I agree on, the clarification will make everyone's life eaiser.

But I repeat. **Do not bundle these two unrelated considerations**, they are not related. As an (occasional) user of low-level `base` stuff, I do care about them being versioned as PVP specifies. Otherwise I'd be *forced* to use tighter upper bounds on `base`. If you want to version GHC specifics separately, put them into separate package. Until then, version them properly.

Finally, if it's really internal modules (i.e. possible future `ghc-base`) maintenance approach is

> The GHC team makes no effort to maintain the stability of this API

then whole thing is not worth doing. You probably want to word that differently.

**Comment by bgamari** [comment]
> If change is breaking, than it requires major version bump.

As a matter of principle, I have no objection to continuing to bump the version of `base` when internal declarations change in a breaking way. This is one reason why I support #145.

However, until this happens (which could happen in time for GHC 9.8, if we move quickly) we will need to be pragmatic (as we have been in the past). Concretely, this means accepting the (small) possibility that breaking changes in internal modules may occur in minor version bumps. Of course, we will do due diligence to minimize the damage, but if there is a change needed in an internal declaration for a soundness fix (or something of similar severity) then we will make the change with a minor bump. Doing otherwise would merely cause undue harm to users for no perceivable benefit.

**Comment by phadej** [comment]
@bgamari

> (as we have been in the past).

Can you provide a short summary of such breaking change deep in `base` which should had technically been major bumps but hadn't? I don't remember any in recent years.

**Comment by bgamari** [comment]
> Can you provide a short summary of such breaking changes deep in base which should had technically been major bumps but hadn't? I don't remember any in recent years.

Such cases are (and should be) rare. Producing a comprehensive summary of these changes would be require a fair amount of effort not because there are numerous cases but because they tend to be small and subtle. On a cursory look through the last few releases I was unable to find a single one.

However, the principle stands: if we need to change, e.g., the type of `GHC.Desugar.toAnnotationWrapper` in a minor GHC (and therefore `base`) release to fix a soundness issue then we should do so without putting the rest of the ecosystem through the pain of a major bump in `base` version. 

To reiterate, ideally `GHC.Desugar` wouldn't live in `base` at all, but sadly we don't yet live in a world where that is true.

**Comment by phadej** [comment]
> To reiterate, ideally the likes of GHC.Desugar wouldn't live in base at all

Does that module needs to be public? If it's used in desugaring, can't GHC access symbols in non-public modules? At least TH can AFAIK.

**Comment by bgamari** [comment]
> EDIT2: Or if it's in non-CLC module, than just doing it and warning users in base changelog, and testing the Stackage that it actually doesn't break anyone.

Yes, this is precisely the pragmatic approach that I was suggesting in the [original comment](https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1474065845).

**Comment by bgamari** [comment]
> `s/divide/classify/`

Quite right. I've made this change.

**Comment by TeofilC** [comment]
I think one thing that might be good to draw out is how we will communicate the status of base modules going forward.

How would a first time contributor to `base` find out if the module they want to modify needs a CLC proposal or not?

Some possibilities:
  - commit a version of Ben's spreadsheet into the CLC repo, link to it from the README and keep it up to date
  - specify the modules status inside the module
  - add a checkbox to the GHC MR template that asks if the user has checked if the change needs a CLC proposal

---

Another question is what the procedure will be for modules moving from one group to another, or for the introduction of new modules. For instance would adding a new internal module need CLC review? What about moving a module from internal to public?

**Comment by bgamari** [comment]
> But that is different from what the proposal says. Especially as proposal says that `GHC.Stats` is internal, meaning that it might change in any minor GHC release. I don't agree on that, such soundness issue might need to wait for major GHC release until `GHC.Stats` is moved to `ghc-base`.

To be clear, this proposal explicitly /does not/ proposal to split out the internal modules of `base` into `ghc-base`. While I believe that doing so would be a reasonable step in the future, the goal of this proposal is merely to introduce a distinction between internal and external modules. We are currently proposing that the status of a module be captured in its documentation.

As long as internal modules remain in `base` we believe that it is reasonable to reserve the right to evolve internal modules outside of the PVP. However, we also agree that the PVP holds value and therefore will naturally take care when exercising this right.

> I think one thing that might be good to draw out is how we will communicate the status of `base` modules going forward.

Yes, clear communication will be essential. I have added a small note suggesting that we introduce a new Haddock field to mark internal modules. Adding a mention to the MR template also seems like a reasonable idea.

> Another question is what the procedure will be for modules moving from one group to another, or for the introduction of new modules. For instance would adding a new internal module need CLC review? What about moving a module from internal to external?

This proposal seeks to establish the /concept/ of an internal module and a roadmap for what modules we would l like to move to internal status in the future. While it would be great if the CLC would summarily accept a swath of these changes, we are willing to propose concrete changes in future, smaller CLC proposals if necessary.

**Comment by hasufell** [comment]
@bgamari are you suggesting this as a stop gap or a permanent solution?

**Comment by bgamari** [comment]
Do note that I have dropped the section entitled "where to place internal declarations" as it contained language from the editing process which was not intended to be in this proposal (namely the renaming of existing modules).

**Comment by hasufell** [comment]
> > @bgamari are you suggesting this as a stop gap or a permanent solution?
> 
> I am suggesting this as a potentially-permanent solution. That being said, I do believe that #145 would be a considerable improvement and would love to see it adopted in the future..

I see.

In that case I'm leaning towards -1 on this proposal, since I believe violating PVP in the standard library sets a bad example (and as indicated in the PVP ticket, I'm also against formalizing it into PVP).

The only pragmatic way forward is the base split to me.

I'm starting to find it hard to follow all the inter-connected tickets about this, though.

**Comment by Bodigrim** [comment]
> For instance, the `GHC.Base.mapFB` function is a necessary exposed part of the fusion framework for `map` but which GHC's authors never intended users to call.

@bgamari could you please elaborate on this? There is no necessity to expose functions used by the list fusion framework, and indeed no other function of it (e. g., `zipWithFB`, `mapAccumLF`, `unwordsFB`) is exposed. It seems `mapFB` is just an oversight and unfit to support your point.

**Comment by bgamari** [comment]
> @bgamari could you please elaborate on this? There is no necessity to expose functions used by the list fusion framework, and indeed no other function of it (e. g., `zipWithFB`, `mapAccumLF`, `unwordsFB`) is exposed. It seems `mapFB` is just an oversight and unfit to support your point.

Yes, this is a fair point; `mapFB` and friends indeed need not be exposed as they are only used in rules defined in `GHC.Base`. However,  even without leaving `GHC.Base` there are other examples; for instance, `maxInt` exists to only serve other modules in `base`.

Looking beyond `GHC.Base`, there are the many exports of `Data.Typeable.Internals` (e.g. `mkTrType`), the array implementation in `GHC.Arr`, and the exports of `GHC.Err`.

**Comment by Bodigrim** [comment]
> However, even without leaving `GHC.Base` there are other examples; for instance, `maxInt` exists to only serve other modules in `base`.

(Assuming for a moment that `maxInt` is potentially unstable)
The only place `maxInt` is used is `GHC.Enum`, it could have been defined there and not exposed outside. Even if it was used in multiple modules, it could have been defined in an `other-module` and never exposed to users. 

More generally, my question is this. I understand that many fragile entities have been exposed from `base`. Is there a genuine reason for them to be publicly exposed, or is it a historical accident / negligence?

**Comment by bgamari** [comment]
> More generally, my question is this. I understand that many fragile entities have been exposed from base. Is there a genuine reason for them to be publicly exposed, or is it a historical accident / negligence?

Yes, there are places where we must expose declarations which otherwise shouldn't be used. Off hand I can think of at least three concrete reasons why this happens:

 * to avoid import cycles (e.g. `GHC.IO.mkUserError`)
 * due to dependencies from generated code (e.g. `Data.Typeable.Internals`)
 * for testing purposes

**Comment by bgamari** [comment]
> More generally, my question is this. I understand that many fragile entities have been exposed from `base`. Is there a genuine reason for them to be publicly exposed (and more specifically - exposed from `base` and not from elsewhere), or is it a historical accident / negligence?

These cases are indeed largely due to historical accident. In many cases modules have been exposed via `exposed-modules` which should have rather been hidden in `other-modules`. These internal modules are often necessary either to serve to break module imports or to satisfy the needs of generated code and were not intended to be used by end-users.

**Comment by Bodigrim** [comment]
Is it difficult to keep modules which have been exposed due to historical accident stable (or at least avoid breaking changes)? And contain new work to `other-modules`?

**Comment by bgamari** [comment]
@Bodigrim it very much depends upon the case; this is essentially what I try to capture in the "stability risk" column of the spreadsheet. Anything assessed to be 0 or 1 will likely be quite easy to keep stable (and consequently I generally propose that these be "stabilized" except in cases where there is no evidence of external usage).

Modules assessed to be a higher stability risk would be harder. In some cases we propose that these be stabilized despite this (in particular, in cases where we find high degrees of dependence in the ecosystem). However, there are certainly a number of modules which we would prefer to hide.

**Comment by Bodigrim** [comment]
What does Stability risk grade 3 stands for?

**Comment by bgamari** [comment]
These assessments are a qualitative, fairly subjective grading. Grading 3 essentially corresponds to things that not only are likely to change but that we would also active discourage users from relying on.

**Comment by bgamari** [comment]
Does anyone else have any thoughts on this? It would be great to be able to move this proposal along.

**Comment by Bodigrim** [comment]
> These assessments are a qualitative, fairly subjective grading. Grading 3 essentially corresponds to things that not only are likely to change but that we would also active discourage users from relying on.

Let's take a look at `System.Posix.Internals`, for example. This is indeed quite a kitchen sink and users would be better off using `unix` package. But I do not see a reason for it to be fundamentally unstable: it's not like we really expect to change the type of `c_open :: CFilePath -> CInt -> CMode -> IO CInt` any time soon. 

Why not put actual implementation into, say, `GHC.Internals.Posix` declared as an `other-module` and keep re-exporting the frozen set of entities from `System.Posix.Internals`? And probably slap `{-# DEPRECATED #-}` over the latter? This way GHC developers gain ability to add new entities to `GHC.Internals.Posix` as they see fit, and PVP is not violated = clients are not disrupted.

My point is that if the majority of exposed internals are historical accidents then we should not perpetuate and deepen these mistakes by excluding them from PVP. Keep them stable, mark as deprecated and put new stuff into `other-modules`. This process can be done module by module, without any big jumps.

**Comment by Bodigrim** [comment]
Based on provided examples, I'm not convinced that it's a good idea to exclude breaking API changes (even only limited to certain modules) from CLC purview. It seems more beneficial to move implementations under `other-modules` and keep already exposed interfaces stable.

I am however quite sympathetic to exclude non-breaking, additive API changes in certain namespaces from CLC process. I'm fine to designate `GHC.Internals` for this. Being mindful that breaking whatever was already released as exposed would require a proposal should help GHC developers to find a right balance in API design on their own, without CLC input.

**Comment by hasufell** [comment]
I'm getting confused by all the proposed variations. So I'll reiterate my stance:

1. base must be PVP compliant (all of it)
2. every module that is exposed in base is under CLC purview

Any proposal that violates those points will get a -1 from me. I'm open to alternatives to base split though.

**Comment by AndreasPK** [comment]
> Based on provided examples, I'm not convinced that it's a good idea to exclude breaking API changes (even only limited to certain modules) from CLC purview. It seems more beneficial to move implementations under other-modules and keep already exposed interfaces stable.

In my experience hiding experts makes certain common tasks like minimizing reproducers for ghc bugs more painful for little benefit so I would not welcome such an approach.

**Comment by nomeata** [comment]
I agree that we should expose all the experts!
If you are an expert, no need to hide, we are all mostly friendly!
(SCNR)

**Comment by simonpj** [comment]
Thanks for the feedback on this proposal.  Ben said above:

> To be clear, this proposal explicitly /does not/ proposal to split out the internal modules of base into ghc-base. While I believe that doing so would be a reasonable step in the future, the goal of this proposal is merely to introduce a distinction between internal and external modules. We are currently proposing that the status of a module be captured in its documentation.

Our intention was to make progress on the idea of identifying which bits of `base` are internal, without having to wait for a machine-supported mechanism to be agreed.  But it's clearly a bit unsatisfactory to have only an informally-defined split, with some special language around the PVP.  @hausfell's puts it well.  He wants the clarity of:

1. base must be PVP compliant (all of it)
2. every module that is exposed in base is under CLC purview

The leading contender for a proper mechanism is, I believe, some version of the "split base" idea. Here is a minimal version:

* We make a new package `ghc-base`, on which `base` depends.
* `ghc-base` is not under the CLC purview.

* Initally, all the code in the current `base` stays there, and `ghc-base` is empty.
* The API of `base` does not change at all.
* GHC devs are free to add new code and exports into `ghc-base`. It is the place to put stuff that is
  * internal to GHC's implementation of the `base` API
  * that we still want to expose to savvy users
  * but which we do not want to expose in the the highly-stable and much-depended-on API of `base`.

* GHC devs are also free to move code from `base` into `ghc-base`, provided the API of `base` does not change. They might need to do so becuase a desired function to add to `ghc-base` depends on something in `base`.  This might mean that modules in `base` become shims that simply re-export stuff from `ghc-base`.

* Over time, anyone can propose to remove modules from the `base` API.  A list of candidates for such removals are the "internal" modules of this proposal.  But there is no big bang. We can do this over time, in batches, whenever.

Questions:
* Is this plan technically feasible?
* Would it enjoy the support of CLC?

Others will know better than I what the technical problems might be.  Someone told me that Haddock doesn't work very well with shim modules... is that right? Is it fixable?

There is plenty of prior art here:
* [GHC ticket #20647](https://gitlab.haskell.org/ghc/ghc/-/issues/20647)
* [HF tech proposal #47](https://github.com/haskellfoundation/tech-proposals/pull/47)
* [CLC proposal #145](https://github.com/haskell/core-libraries-committee/issues/145)

Everyone is trying to do the right thing here.  I'm keen for us to find a way forward
that might not do *everything*, but does *enough* to allow us to make progress.

**Comment by mixphix** [comment]
Thanks for this great summary. This use of `ghc-base` will allow GHC developers the freedom to hack and experiment as they see fit without as great a risk of breaking the rest of the world.

When moving modules from `base` into `ghc-base`, we will have to be certain not to touch the export lists, or to make explicit (in all the exported names, not just the module itself) those that do not have them: this *should* prevent Haddock from rendering a blank module.

**Comment by hasufell** [comment]
I think what @simonpj describes sounds somewhat similar to what @Bodigrim wrote here: https://github.com/haskell/core-libraries-committee/issues/145#issuecomment-1484182300

However, still unanswered is the question who will govern `ghc-base`. Because as pointed out: if base re-exports from ghc-base, CLC must govern both (or any function/type that is a transitive dependency of any base exported function... who will keep track of this subset of ghc-base?). Otherwise it can't fulfill its duty.

To make it more clear. Assuming we have a consensus on what "GHC internals" means, we have

1. base API that doesn't depend on GHC internals
2. base API that does depend on GHC internals
3. GHC internals that depend on base API
4. GHC internals that don't depend on base API

The problem is 2. and 3, because both would have to be moved into `ghc-base` eventually, if we want all GHC internals to live there. Even if we go with the gradual approach, the only thing that's feasible without major effort and communication is migrating 4.

However, even if we stagnate with 4., it might still be an improvement and allow to deprecate some of the GHC internals in base.

**Comment by chshersh** [comment]
I propose to start with `ghc-base` that depends on `base`. We already have "`ghc-base` used by `base`" and it's called `ghc-prim`.

**Comment by mixphix** [comment]
That's a good point, we already do have `ghc-prim` for this purpose. I'd hate for the GHC codebase to become even more convoluted and unmanageable than it already is (or appears to be). Limiting the interwovenness of internal packages and organizing core code in a sensible way will also improve stability, ease of onboarding, and community participation.

**Comment by simonpj** [comment]
> However, still unanswered is the question who will govern ghc-base. Because as pointed out: if base re-exports from ghc-base, CLC must govern both (or any function/type that is a transitive dependency of any base exported function... who will keep track of this subset of ghc-base?). Otherwise CLC can't fulfill its duty.

I think the answer is simple:

* GHC devs control `ghc-base`
* But if any change there is visible in the API of `base`, we must consult CLC

Simple!  CLC controls the API of `base`, and changes there need their approval.

I think `ghc-prim` is a bit of a red herring here.  It exists as a separate package so that we can build the integer package, on which `base` depends.  It should not matter to clients how GHC structures its internals -- it's an implementation matter.  Of course, we have a strong incentive to make it as simple as possible!

To put it another way, GHC could implement `ghc-base` today, and provided we don't change the API of `base`, no one outside GHC-dev-land should know or care.  I'm just floating the idea here because I think we'll all work together better if we commmunicate well.   And I'm a little hazy about whether our tech (esp Haddock) is up ot making it invisible whether a function is implemented in `base` itself or is a re-export of something from another package.  (Of course this already arises for types and functions defined in `ghc-prim`.)

**Comment by hasufell** [comment]
> But if any change there is visible in the API of base, we must consult CLC

This isn't always that straight forward in light of transitive function dependencies (e.g. function A is re-exported from base, but also uses internal function B, which is not re-exported... GHC devs change function B and accidentally break performance properties).

Remember, CLC does cares about laziness/performance properties.

Who is going to asses whether there is a visible impact on base API?

**Comment by bgamari** [comment]
@hasufell, I think that the transitive dependency issue is surmountable. GHC developers, aided by CI, would assess impact. We already have work-in-progress to verify that `base`'s exports and their types are not inadvertently changed. In principle we could (and perhaps should) add tests to verify the laziness properties of its exports. Performance is clearly harder, but I would argue that the difficulty imposed by splitting `base` and `ghc-base` is negligible.

**Comment by simonpj** [comment]
> Who is going to assess whether there is a visible impact on base API?

Good question.  @bgamari has a patch that mechanically checks the API of `base`, and complains if there are any changes. This is way more than we have right now!

> This isn't always that straight forward in light of transitive function dependencies (e.g. function A is re-exported from base, but also uses internal function B, which is not re-exported... GHC devs change function B and accidentally break performance properties).

That is true, but is *already* a challenge:
* `base` depends on `ghc-bignum` and `ghc-prim`, so changes in either of those packages could change performance.
* I don't think CLC currently reviews *every* patch to `base`, only those that the author thinks might affect performance.  It'd be the same here.  Changes to `ghc-base` that affect performance of `base` functions shoudl be discussed with CLC.
* Every patch to GHC itself can affect performance characteristics of compiled code, occasionally substantially.

One Good Thing would be to expand the performance tests in the `base` test suite, and that is something the CLC might consider.  But I see no systematic way to *guarantee* that a changes to GHC or its ecosystem of libaries won't have an effect on the perf of one or more `base` exports.  We just all have to do the best we can -- and that "best" is unaffected by whether the code happens to live in `base` or in `ghc-base`.

Another bit of technology is on its way: deprecated exports.  [GHC proposal #134](https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0134-deprecating-exports-proposal.rst), [GHC ticket #4879](https://gitlab.haskell.org/ghc/ghc/-/issues/4879), and an up-coming merge request. This will make it easier when we went to remove internal modules from `base` and export them from `ghc-base` only.

Apparently the Haddock issue really isn't an issue.  For example, [here is GHC.Exts](https://hackage.haskell.org/package/base-4.18.0.0/docs/GHC-Exts.html) which is mostly just a shim, and its Haddock page looks just fine.

**Comment by simonpj** [comment]
> Please share you opinion on "split base" in the relevant issue

You mean [here](https://github.com/haskellfoundation/tech-proposals/pull/47)?  Yes, good idea.  I will do that.  My reason for floating it here is that, given the lack of enthusiam for our proposal above I wanted to gauge the CLC enthusiasm for:
* Withdraw our proposal above
* Support the split-base proposal
* Then re-make our proposal in the form of proposals to remove GHC-internal modules from the `base` API and put them in `ghc-base`

Question: would that plan meet with CLC approval? I am keen to pursue a path that enjoys CLC support.

> There seems to be an underlying notion that asking for CLC approval is a burden and a bottleneck

Not at all.  It's rather that I want a clear way to signal *to users* that a particular data type or function is internal, and should only be used if you want to be exposed to GHC internals.  We have no way to signal that at the moment, and that inevitably leads users to accidentally depend on things they shouldn't, through no fault of their own; and then pressure to keep these accidental and unintended interfaces stable into future.   By providing a structural way to express intent, we will avoid future pain.

**Comment by mixphix** [comment]
Forgive me if this is a silly question, but do the `exposed-modules`/`other-modules` distinction in the .cabal file not provide enough distinction? If these are indeed cabal packages, there is support for "internal libraries" as well. This would allow an equally-precise segregation of "within CLC scope" modules and "outside CLC scope" modules, without stumbling into the larger issue of actually cleaving `base` in two.

**Comment by simonpj** [comment]
> do the exposed-modules/other-modules distinction in the .cabal file not provide enough distinction?

Alas, no.  `other-modules` cannot be imported, ever.  But high-performance users specifically want to reach into the internals of GHC (e.g. the representation of arbitrary precision integers) even at the cost of being exposed to churn in those internals.  Performance testing and debugging are other reasons that we want it to be possible to import internal modules.  If cabal provided a 3-way distinction (hidden, internal, external) with the latter two being exposed, that would indeed address the problem. 

But changing cabal would take time, and some people say "no need to change cabal, just split your package into two", so it's not a slam-dunk change.

**Comment by Kleidukos** [comment]
(With the new feature of multiple public library components in cabal, this split could be made easier to manage)

**Comment by hasufell** [comment]
Let's go back to this very proposal:

- declaring GHC modules as internal makes sense to me... but this doesn't really need CLC approval anyway
- violating PVP is not cool, so I'm against this part
- excluding modules from CLC purview (given that they adhere to PVP)... I could agree to, if we make it very clear how API impact is going to be evaluated and communicated 

If we agree to internal modules, excluded from CLC purview, while maintaining PVP adherence... then that doesn't seem too bad. Although one could argue it might cause churn, because of more aggressive bumping of the major base version... I don't think it matters much, since GHC ships with base anyway (although theoretically, a GHC patch release could introduce a major base bump, which can be problematic... we might just have to avoid it).

In case we want base to be reinstallable and decouple it from GHC, then we can execute the base split. Because then, base PVP is much more useful/important than it is now.

So that could be a possible stopgap.

**Comment by Bodigrim** [comment]
My analysis is similar to @hasufell here. 

* I'm uneasy with **explicitly** granting PVP exclusions. Given that the proposer (or anyone else) cannot recall a use case, I'd rather omit this for now. FWIW CLC does not normally opine on versioning of `base`, GHC developers are bumping versions on their own.

* I'm fine with documenting some (potentially, a wide range of) modules as internal. In my view this is a necessary prerequisite for a potential future package split into `ghc-base` and `base`, if any.

* I'm generally agreeable with allowing "additive" changes to "internal" modules, potentially even without CLC approval. There is a certain subtlety about who is to decide that the change is purely additive. There is a usual complication with type classes and instances: adding a new type class or instance to `base` could have far-fetching consequences, so it might be worth a consultation. 

> > There seems to be an underlying notion that asking for CLC approval is a burden and a bottleneck
> 
> Not at all.

Interesting. I'd say that excluding modules from CLC purview is the most contentious part of the proposal. If it is not a key point, I'd suggest (in my private capacity) the following:

* We document and designate a range of modules and namespaces in `base` (roughly in line with Ben's analysis) as "internal". 
* We agree that the bar for changes to "internal" modules is generally lower than to "non-internal" ones. They nevertheless remain subject to PVP and require a CLC proposal.
* We agree that additive changes to "internal" modules are generally acceptable and can be approved following a lightweight procedure. We also agree that uncontrolled expansion of `base`, even in "internal" parts, is not in our best interest.
* We do not sneak stuff, which is generally useful to wide circles of users, into "internal" modules. E. g., it's unfair to put `class Profunctor` into, say, `GHC.Internal.Profunctor` module, because everyone will ignore warnings about instability and just go for it.

@simonpj @bgamari does it look reasonable to you?

**Comment by cdornan** [comment]
I am a little confused here. Ben is proposing to provide some structure where there is currently none (in effect) so that the GHC team and the CLC can do a better job of managing change, and so that everyone else can better understand what is stable and can be relied upon, and what is internal and more liable to change. This is a good, yes?

Does it make sense to object on the grounds that any change anywhere might be visible and therefore every change whatsoever going forward to anything in GHC must be subjected to a CLC approval process? This seems to be the logic of the objections, applied consistently.

Can we not make things a little better without trying to make them perfect. For somebody outside of GHC and the CLC Simon's proposal appears perfectly resonable.

**Comment by cdornan** [comment]
@bodigrim's proposal seems reasonable to me but the devil could be in the detail. I will be interested to hear what Ben and Simon think.

**Comment by adamgundry** [comment]
A related point that would be good to clarify is the case of modules added or modified following the GHC proposals process. There are frequently cases where a GHC proposal requires coordinated changes to the compiler and `base` (certainly as long as we do not have `ghc-base`). Some examples that come to mind are:
 * https://github.com/ghc-proposals/ghc-proposals/pull/562 which proposes changes to `Symbol` and goes to some lengths to avoid specifying a public API for it so that the public API can be left for CLC (but then is running into problems justifying the change at all).
 * https://github.com/ghc-proposals/ghc-proposals/pull/330 which proposes changes to public APIs related to exception handling.
 * #162 which asks for CLC approval for changes that the GHC steering committee has previously approved.

I'm particularly concerned that we may discourage contributors (both new and existing) if the process for getting changes through is overly bureaucratic.

**Comment by simonpj** [comment]
> A related point that would be good to clarify is the case of modules added or modified following the GHC proposals process. There are frequently cases where a GHC proposal requires coordinated changes to the compiler and base (certainly as long as we do not have ghc-base).

Yes I agree.  But @Bodigrim, rightly, doesn't like conflating multiple conversations on the same issue.  

So where would be a good place to have this conversation?
* Right here in this issue?
* In a new CLC issue (with participation from the GHC Steering Committee)?
* In a GHC Proposals issue (with participation from CLC)?

@Bodigrim if you say, we'll follow your guidance.

**Comment by simonpj** [comment]
@Bodigrim suggests
*  We document and designate a range of modules and namespaces in base (roughly in line with Ben's analysis) as "internal".
* We agree that the bar for changes to "internal" modules is generally lower than to "non-internal" ones. They nevertheless remain subject to PVP and require a CLC proposal.
* We agree that additive changes to "internal" modules are generally acceptable and can be approved following a lightweight procedure. We also agree that uncontrolled expansion of base, even in "internal" parts, is not in our best interest.
* We do not sneak stuff, which is generally useful to wide circles of users, into "internal" modules. E. g., it's unfair to put class Profunctor into, say, GHC.Internal.Profunctor module, because everyone will ignore warnings about instability and just go for it.

That sounds like a good way forward to me!  It would allo us to
* Signal (albeit informally) to users which modules are "internal" in the sense dicussed above.
* Allow us to get on with discuss which modules are "internal" and which are not.

But I would like to add:
* Let's pursue the split-base plan as well.  Once it is in place we can:
  * Release experimental APIs via `ghc-base` and later migrate them into `base`
  * Make proposals to CLC (which can be discussed later) to move modules agreed as internal from `base` into `ghc-base`, accompanied with suitable deprecation periods etc.

Great!   Would the CLC be content with this path?

**Comment by bgamari** [comment]
For what it's worth, I have recently stumbled across a use-case where we would need to expose internal modules from `base`. Specifically, the [exception backtrace proposal]() proposes that exceptions carry stack backtrace information represented by GHC's new `StackSnapshot#` type, among other representations. It further proposes that backtraces can be decoded and formatted for display to the user. The logic for decoding such stack snapshots currently lives in the `ghc-heap` package, which cannot be depended upon by `base` (since this would induce a dependency cycle).

One way to address this would be to move the implementation of `ghc-heap` into `base` and expose it as a set of internal modules. These could then be used by `base` and re-exported from `ghc-heap` (which would continue to be their canonical home).

**Comment by phadej** [comment]
> One way to address this would be to move the implementation of `ghc-heap` into `base`

... or just keep the decoding in `ghc-heap`. And when `ghc-base` emerges move stuff there then. And you won't need to involve CLC at all at any point.

**Comment by bgamari** [comment]
> ... or just keep the decoding in `ghc-heap`. And when `ghc-base` emerges move stuff there then. And you won't need to involve CLC at all at any point.

Yes, that would be another option. However, my comment was intended to address @Bodigrim's [previous comment](https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1500865167) suggesting that `other-modules` was adequate to hide internal implementation details. At the time I didn't have a good example of why this was not sufficient; this comment was intended to provide such a scenario.

As a practical matter, I do worry that predicating stack decoding support on `ghc-base` would forestall proper backtrace support in GHC for at least one release cycle.

**Comment by phadej** [comment]
> for at least one release cycle.

That's would been an argument if release cycle would been a year. But the release cycle is supposed to be six months, and one reason it's six months, if i remember correctly, is so that there isn't pressure to try to crank everything possible into the next release, delaying the release.

Nevertheless 

> I do worry that predicating stack decoding support on `ghc-base`. 

`ghc-heap` is a fine place for decoding for people who need that, so no need to predicate on `ghc-base` done first. `StackSnapshot#` sounds like a very GHC specific thing.

If you'd move `ghc-heap` into `base` (even "internal") modules, you'll hard force `base` major bump for every GHC release, because AFAICS, `ghc-heap` had always had major bumps with GHC releases (and it doesn't have changelog, so it's hard to say whether there were actual breaking changes, but the safe assumption that there are, as GHC heap stuff do change).

That would be a backwards development, if the long term goal is to decouple `base` versions from GHC versions.

TL;DR I see that as a bad example. The solution you propose is "easy", but step into the wrong direction.

**Comment by bgamari** [comment]
> If you'd move `ghc-heap` into `base` (even "internal") modules, you'll hard force `base` major bump for every GHC release, because AFAICS, `ghc-heap` had always had major bumps with GHC releases (and it doesn't have changelog, so it's hard to say whether there were actual breaking changes, but the safe assumption that there are, as GHC heap stuff do change).

I think my comment may have been unclear here: I am proposing this as an example of why having a set of internal  modules exposed by `base` (but not covered by the PVP) may be useful. The point of designating these as "internal" modules is specifically so we would not need to bump `base`'s version number in response to changes. @Bodigrim had [previously](https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1532251480) said he was uneasy with granting explicit PVP exemptions, in part because we could provide no concrete need. My goal here is to demonstrate one such possible need.

Of course, I agree that `ghc-base` is a better design. I will leave an additional comment on #145 as I feel this `ghc-heap` example serves as equally good motivation for the `ghc-base` change as it does for the internal module change proposed here.

**Comment by phadej** [comment]
> (but not covered by the PVP)

I disagree with this need to begin with.

Code bundled with GHC has to be versioned properly. All of it. Whole ecosystem builds on top of it. If library code changes in sync with GHC version, version the library in sync with GHC.

It is additional work at the GHC end (= you cannot "just" make changes), but everything else would benefit from that.

**Comment by bgamari** [comment]
> Code bundled with GHC has to be versioned properly. All of it. Whole ecosystem builds on top of it. If library code changes in sync with GHC version, version the library in sync with GHC.

My objection to this centers on the phrase "all of it". The original intent of this issue was to make explicit the fact that there are portions of `base`'s surface area which either:
 * must be exposed but we do not believe users should be depending upon, or
 * we expose solely for users who are willing to put in a bit more work to exploit internal, potentially unstable properties of GHC's implementation; in such cases the onus is on the user to keep up with upstream changes.

Again, ultimately the goal of my bringing up this `ghc-heap` issue was to demonstrate there are concrete cases where `other-modules` is insufficient and where PVP compliance is unduly burdensome. However, I am also very happy to instead pursue the `ghc-base` option. Ultimately, however, I strongly believe that we need one of these options; under the status quo we are simply stuck without any way forward.

**Comment by simonpj** [comment]
I [asked above](https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1535470894)

> Great! Would the CLC be content with this path?

I think @Bodigrim is on a well deserved holiday, but I'd love to hear the opinions of other members of the CLC.

**Comment by Bodigrim** [comment]
> But I would like to add:
> 
>   * Let's pursue the split-base plan as well.  

That would be inappropriate: there is a separate proposal to discuss `base` split in full detail. 

This meta-proposal is fundamentally about working arrangements between GHC developers and CLC. We can agree a (potentially underspecified or handwavy) policy, see how it goes and come up with amends or reversal later. And indeed what I suggested in https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1532251480 is more of a framework for collaboration in good faith rather than set-in-stone rules to police.

But `base` split is a one-way road, we would not be able to revert it once enacted, which makes it fundamentally different from the topic of this proposal. We should not make irreversible decisions without a proper and detailed proposal, and there is a dedicated space to develop it.

Frankly speaking, Simon, I find your insistance to use every opportunity to turn this discussion into `base` split track bizarre. We already discussed [above](https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1530079515) that it does not belong here. 

----

I've been told by other CLC members that they find this discussion difficult to follow, and I can relate to this problem myself. As a person responsible to see proposals through, I'd like to urge all participants to stick to the topic closely and refrain from tangential remarks (although feel free to open new issues). 

I'm sorry if you find this call offensive, but CLC members are all unpaid volunteers, and limiting hot discussions to their immediate topic is vital to get anything done in a very limited amount of time we can dedicate.

**Comment by mixphix** [comment]
Thanks @Bodigrim for doing your absolute best to follow this thread (and others) related to the `ghc-base` split. I admit that the never-ending tangents and vagueness about what the GHC developers plan to do with their codebase has exhausted my ability to pay attention to this issue. Frankly, I find that the lack of coordination in this thread (and others) is indicative of deeper problems in the management of the development of the compiler that lie outside the CLC’s scope of operation. When the GHC team can propose a detailed, concrete solution for their own codebase, the CLC will be able to provide a detailed, concrete response. Until then, I maintain the firm position that `base` ought not to be meddled with by those who would destabilize the entire Haskell ecosystem with unstable changes.

**Comment by hasufell** [comment]
> > (but not covered by the PVP)
> 
> I disagree with this need to begin with.
> 
> Code bundled with GHC has to be versioned properly. All of it. Whole ecosystem builds on top of it. If library code changes in sync with GHC version, version the library in sync with GHC.
> 
> It is additional work at the GHC end (= you cannot "just" make changes everywhere at will), but everything else would benefit from that additional work.
> 

I strongly agree. I can't understand why GHC HQ is so eager finding exemptions to PVP.

> and where PVP compliance is unduly burdensome

I question that this is the case and I wonder if this is maybe an issue with GHC development practices. But I wouldn't know, since I've never written a compiler.

However, non-compliance is incredibly burdensome to API consumers. So I think it's rather fair that this burden lies at the hands of us (CLC, GHC, boot library maintainers).

**Comment by cdornan** [comment]
I remain perplexed. The intention here is to separate base into:

  * 'public' parts that form part of the public API and are subject to PVP conventions;
  * 'internal' parts without the same stability contract, that folks can reach into if they really need to (but it is not expected that most developers will need access to these internals).
 
This is a common pattern adopted by package maintainers and surely an improvement on the current situation where GHC changes manifest in ways that are not reflected in the API origanisation, because, in this respect, there isn't any.

Would it help to discuss this on a call?

**Comment by hasufell** [comment]
> This is a common pattern adopted by package maintainers

That doesn't mean it's a good pattern.

There is also lots of packages on hackage that don't maintain their version bounds properly (and e.g. assume people use stack resolvers or cabal freeze files). Those have to be cleaned up by hackage trustees at times and causes work for many parties.

I don't think base should follow bad examples.

E.g. `bytestring` exposes internals, but puts huge effort into maintaining compatibility (e.g. through pattern synonyms).

We know there's a **proper** pattern for exposing internals in a PVP compliant way: https://nikita-volkov.github.io/internal-convention-is-a-mistake/

Since the above pattern is not discussed as a solution in this ticket, the conclusion is to, well, just be PVP compliant with Internal modules. I don't even find evidence why that would currently be a major problem.

**Comment by simonpj** [comment]
My apologies @Bodigrim.   I was trying to see if a consensus existed for a way ahead, but I will take your advice and pursue that exclusively on the [HF tech-track proposal #47](https://github.com/haskellfoundation/tech-proposals/pull/47)

Returning to this thread, then, are we agreed to follow your plan, which I reproduce here? 

1. We document and designate a range of modules and namespaces in base (roughly in line with Ben's analysis) as "internal".
2. We agree that the bar for changes to "internal" modules is generally lower than to "non-internal" ones. They nevertheless remain subject to PVP and require a CLC proposal.
3. We agree that additive changes to "internal" modules are generally acceptable and can be approved following a lightweight procedure. We also agree that uncontrolled expansion of base, even in "internal" parts, is not in our best interest.
4. We do not sneak stuff, which is generally useful to wide circles of users, into "internal" modules. E. g., it's unfair to put class Profunctor into, say, GHC.Internal.Profunctor module, because everyone will ignore warnings about instability and just go for it.

Concerning (1), is [Ben's spreadsheet](https://docs.google.com/spreadsheets/d/1WmyYLbJIMk9Q-vK4No5qvKIIdIZwhhFFlw6iVWd1xNQ/edit#gid=1315971213) adequate designation?

**Comment by mixphix** [comment]
The spreadsheet goes a long way to clarifying the effect of this proposal, but is still very wishy-washy about the end result: many of the modules are marked “stabilize?” as if you are asking us for reassurance, and I see no indication of a resulting module tree for us to examine once these changes have been made, despite several comments suggesting to rename or re-export things from modules that do not yet exist.

I am against introducing a “lightweight” version of the proposal process even for purely additive changes to these internal modules. Well-argued and well-pursued proposals have generally been approved with little (community or CLC) pushback, or have been found through productive discussion to be antithetical to the sum of our individual ideas on how the library should evolve. In my view, changes by compiler developers that affect the base library (and transitively, the entire Haskell ecosystem) *must* be well-argued and well-pursued.

I’m sure writing CLC proposals is far less entertaining than bug-squashing and feature-adding. Yet it remains a vital step in assuring the safety and stability of external code. Suffice it to say that GHC, while being an excellent place to conduct compiler research, is also subject to the needs and demands of the broader software industry. We now we have the organizational structure necessary to ensure that “Code Health and Safety” measures are followed properly. Cease and desist from undermining it.

**Comment by tomjaguarpaw** [comment]
Can we all please pause to check we're taking due care in the language we use in these discussions? Ultimately all parties have the same goal: to continue to improve the ecosystem of the language we love. I think we should default to assumptions of good faith all round.

**Comment by bgamari** [comment]
@Bodigrim [asks](https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1532251480),

> @simonpj @bgamari does it look reasonable to you?

To clarify, @simonpj speaks for me when he says that this compromise looks acceptable. I also agree that we should continue to discuss a `ghc-base` split, but this is indeed an orthogonal matter which we can discuss elsewhere.

**Comment by mixphix** [comment]
Thank you for the clarifications!

I have no issues with the new module stability descriptions. What I would like to see before proceeding is an outline of the new modules and their planned export lists, so that we may make announcements far ahead of time on relevant platforms to ensure the smoothest transition to this new organizational structure.

**Comment by bgamari** [comment]
@mixphix, the plan would be to freeze the export lists of all existing exposed modules at their state as of a designated GHC release (e.g. 9.6.1). As for new modules, I propose the following:

```haskell
module Data.Enum where
  type Bounded :: * -> Constraint
  class Bounded a
    ...
  type Enum :: * -> Constraint
  class Enum a
  enumFrom :: Enum a => a -> [a]
  enumFromThen :: Enum a => a -> a -> [a]
  enumFromThenTo :: Enum a => a -> a -> a -> [a]
  enumFromTo :: Enum a => a -> a -> [a]
  fromEnum :: Enum a => a -> GHC.Types.Int
  maxBound :: Bounded a => a
  minBound :: Bounded a => a
  pred :: Enum a => a -> a
  succ :: Enum a => a -> a
  toEnum :: Enum a => GHC.Types.Int -> a

module Data.Show where
  -- Safety: Safe-Inferred
  type Show :: * -> Constraint
  class Show a
    ...
  type ShowS :: *
  type ShowS = GHC.Base.String -> GHC.Base.String
  show :: Show a => a -> GHC.Base.String
  showChar :: GHC.Types.Char -> ShowS
  showCommaSpace :: ShowS
  showList :: Show a => [a] -> ShowS
  showLitChar :: GHC.Types.Char -> ShowS
  showLitString :: GHC.Base.String -> ShowS
  showMultiLineString :: GHC.Base.String -> [GHC.Base.String]
  showParen :: GHC.Types.Bool -> ShowS -> ShowS
  showSpace :: ShowS
  showString :: GHC.Base.String -> ShowS
  shows :: Show a => a -> ShowS
  showsPrec :: Show a => GHC.Types.Int -> a -> ShowS
```
In addition, there are a few internal functions (confirmed to have no users on Hackage) which I did not export.
A full accounting of the changes being proposed can be found [here](https://gitlab.haskell.org/ghc/ghc/-/compare/wip%2Fci-interface-stability...wip%2Fbase-stability?from_project_id=1&straight=false#2ca7b1d6bb24cd05ca30a428efb924268c8dbc51).

**Comment by Bodigrim** [comment]
I imagine the material part of the proposal would be to add the following disclaimer to the modules with stability risk 2 and 3.

> The API of this module is unstable and not meant to be consumed by general public. If you absolutely **must** depend on it, make sure to use a tight upper bound, e. g., `base < 4.X`, not just `base < 5`, because the interface can change rapidly without much warning.

I'm carefully avoiding word "internal" here because it is too overloaded: after all most of these modules are already marked as internal in haddock metadata. The modules remain subject to usual PVP rules and are not to be disrupted within the same major version of `base-4.X`. 

These are modules with stability risk >= 2 according to Ben's spreadsheet:

* `GHC.Arr`
* `GHC.Conc`
* `GHC.Conc.IO`
* `GHC.Encoding.UTF8`
* `GHC.Exception`
* `GHC.Exception.Type`
* `GHC.ExecutionStack`
* `GHC.Fingerprint.Type`
* `GHC.InfoProv`
* `GHC.IO`
* `GHC.IO.Buffer`
* `GHC.IO.Device`
* `GHC.IO.Encoding`
* `GHC.IO.Exception`
* `GHC.IO.Handle.Text`
* `GHC.RTS.Flags`
* `GHC.Stack.Types`
* `GHC.Stats`
* `GHC.TopHandler`
* `GHC.TypeLits.Internal`
* `GHC.TypeNats.Internal`
* `GHC.Desugar`
* `GHC.Event.TimeOut`
* `GHC.ExecutionStack.Internal`
* `GHC.Float.RealFracMethods`
* `GHC.GHCi`
* `GHC.GHCi.Helpers`
* `GHC.IO.Handle.Internals`
* `GHC.IO.Handle.Types`
* `GHC.IO.SubSystem`
* `GHC.IOPort`
* `System.Posix.Internals`

Once the disclaimer is in place, CLC will be able to exercise less caution when evaluating changes to the listed modules, hopefully reducing friction.

@simonpj @bgamari does it sound right?

Dear CLC members, would such arrangement work for you?

**Comment by simonpj** [comment]
> @simonpj @bgamari does it sound right?

Yes, to me that sounds helpful, thanks.

**Comment by tomjaguarpaw** [comment]
> Dear CLC members, would such arrangement work for you?

Yes, I think the first step of clearly marking as unstable modules that we already know to be so is long overdue.

**Comment by mixphix** [comment]
> > Dear CLC members, would such arrangement work for you?

Works for me.

> Yes, I think the first step of clearly marking as unstable modules that we already know to be so is long overdue.

Not only marking unstable modules as such, but also clarifying the rigidity of the ones that have up to now been marked "unstable" despite no evidence and widespread use.

**Comment by hasufell** [comment]
> > Dear CLC members, would such arrangement work for you?
> 
> Yes, I think the first step of clearly marking as unstable modules that we already know to be so is long overdue.
> 

It's not just about communication to the user, but also whether we'll be ok with being less strict on e.g. API breaking changes. You haven't commented on that part.

**Comment by tomjaguarpaw** [comment]
> It's not just about communication to the user, but also whether we'll be ok with being less strict on e.g. API breaking changes. You haven't commented on that part.

I didn't see that part as changing anything substantive. @Bodigrim said

> CLC will be able to exercise less caution when evaluating changes to the listed modules, hopefully reducing friction.

and to me that sounds like each member of CLC is allowed (but not required) to exercise less caution at their own discretion (which is not really a change from the status quo).

**Comment by hasufell** [comment]
> and to me that sounds like each member of CLC is allowed (but not required) to exercise less caution at their own discretion (which is not really a change from the status quo).

Sure, but in order for this arrangement to be working in practice, CLC members should indicate whether they're ok with lowering their standards for the modules in question or whether they intend to treat it as strict as the rest.

If half of CLC doesn't intend to be more permissive, we may have to reevaluate.

**Comment by tomjaguarpaw** [comment]
Currently my approach is to judge every proposal on a case-by-case basis taking into account the precedent of previous decisions, so I don't expect anything to change from my point of view, vis-a-vis the decision-making process. That is to say, if this proposal is passed it sets the precedent that CLC members ought to consider those modules as less stable, and I'm happy to take that precedent as a strong indicator of how I ought to make decisions.

**Comment by chshersh** [comment]
> Sure, but in order for this arrangement to be working in practice, CLC members should indicate whether they're ok with lowering their standards for the modules in question or whether they intend to treat it as strict as the rest.

I'm okay with lowering my standards if it helps Haskell and GHC users to be more productive.

In fact, I already didn't have high requirements for changes in the GHC-internal (as I see) modules. Formalizing this convention and having a list of such modules explicit and uniform for everyone sounds like a step forward.

**Comment by tomjaguarpaw** [comment]
+1

**Comment by hasufell** [comment]
+1

**Comment by chshersh** [comment]
+1

**Comment by parsonsmatt** [comment]
+1

**Comment by bgamari** [comment]
@Bodigrim thanks for the ping. I have not forgotten; to the contrary, we have been hard at work getting the groundwork in place to implement this and the rest of our proposal.

**Comment by chreekat** [comment]
Could someone perhaps link to (or copy outright) the content of https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1591871779 in the issue description, which is what actually ended up being voted on for this issue? :)  This thread goes for a wild ride that perhaps most future readers don't need to go through. In the end the proposed change was literally just documentation, if I'm not mistaken, while *many* other things of far greater import were argued for and against, as everyone gradually came to understand each other over time.

**Comment by simonpj** [comment]
> @bgamari this is a gentle reminder to prepare an MR implementing the proposal.

> @Bodigrim thanks for the ping. I have not forgotten; to the contrary, we have been hard at work getting the groundwork in place to implement this and the rest of our proposal.

> Could someone perhaps link to (or copy outright) the content of https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1591871779 into the issue description, since it is what actually ended up being voted on for this issue?

I think we are waiting for @bgamari to formulate a concrete proposal (indeed along the lines of the comment you cite).  

Ben, it may well be better to start a new issue, to avoid the "wild ride".

**Comment by Bodigrim** [comment]
> I think we are waiting for @bgamari to formulate a concrete proposal (indeed along the lines of the comment you cite).
>
> Ben, it may well be better to start a new issue, to avoid the "wild ride".

I'm somehwhat confused. There was a very specific proposal voted on in this thread; it's an MR which is missing, but no reason to start a new issue for it.

**Comment by bgamari** [comment]
Yes, I am also rather confused. My understanding is that the proposal as-written in the summary is accepted. The operative changes are described in https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1544775797 and https://github.com/haskell/core-libraries-committee/issues/146#issuecomment-1579347317.

However, I now realize that I did forget to notify the CLC of the implementation of the proposal, which was carried out in [!10422](https://gitlab.haskell.org/ghc/ghc/-/merge_requests/10422). This introduced `Data.Enum` and added the appropriate disclaimers.  Over time I expect that we will open further proposals to deprecate various exports of the modules covered by the proposal.

**Comment by Bodigrim** [comment]
Ben, now I'm confused even more. The approved changes we voted on are in the second comment you linked only, not the first one. I thought I asked both you and Simon, and repeated again the precise scope in the call for voting. If you want to introduce `Data.Enum` or anything else, I'd suggest raising a separate proposal. 

(I've been very slow recently and accumulated a huge backlog of work here, and I'm sorry for this. I hope to clear up things during the upcoming months)

**Comment by bgamari** [comment]
> Ben, now I'm confused even more. The approved changes we voted on are in the second comment you linked only, not the first one. I thought I asked both you and Simon, and repeated again the precise scope in the call for voting. If you want to introduce Data.Enum or anything else, I'd suggest raising a separate proposal.

Apologies, this was my miscommunication. I will raise a proposal for `Data.Enum`.

> (I've been very slow recently and accumulated a huge backlog of work here, and I'm sorry for this. I hope to clear up things during the upcoming months)

Quite alright. You have a lot on your plate.

**Comment by chreekat** [comment]
Thanks for the clarification, @Bodigrim and @bgamari !
